### PR TITLE
feat: multipart openapi

### PR DESCRIPTION
**Pull request by aarnphm**
added openapi supports for IO

also I added pandas and stubs :(

This could mainly due to setup issues. However, for consistency and doesn't let these problems fail CI for future contributor, I think it would be best if we keep a copy of stubs for the library. (applies to library that doesn't have py.typed supports, pandas).

Another solution is to have the correct setup, in which I don't really know how to make it work with python management system like pyenv or pipenv

image

### CODE REVIEW

**Comment by aarnphm** [comment]
codestyle check will definitely fail because the stubs are incomplete. I think we can work the stubs along the way

**Comment by bojiang** [comment]
https://docs.github.com/en/repositories/working-with-files/managing-files/customizing-how-changed-files-appear-on-github

let's just collapse those generated files automatically

File: bentoml/_internal/io_descriptors/base.py [hunk 1]
```diff
@@ -7,6 +8,16 @@
 IOPyObj = t.TypeVar("IOPyObj")
 
 
+def _mk_repr(obj: t.Any) -> str:
```
Comment (bojiang) [inline bentoml/_internal/io_descriptors/base.py:11]: `__repr__` goal is to be unambiguous. Mostly you can eval a repr to create the same object.
`__str__` goal is to be readable.

File: bentoml/_internal/io_descriptors/json.py [hunk 2]
```diff
@@ -123,30 +124,31 @@ def __init__(
...
     async def from_http_request(self, request: Request) -> JSONType:
         json_obj = await request.json()
         if hasattr(self, "_pydantic_model") and self._validate_json:
+            if pydantic is None:
+                raise MissingDependencyException(
+                    "`pydantic` must be installed to use `pydantic_model`"
+                )
             try:
-                return self._pydantic_model.parse_obj(json_obj)
+                self._pydantic_model.parse_obj(json_obj)
```
Comment (bojiang) [inline bentoml/_internal/io_descriptors/json.py:148]: So just verifying but not regulating the json_obj right?

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/json.py:148]: `parse_obj` actually return a `pydantic.BaseModel` instead of a JSON object. I think we should only use pydantic for validation

File: bentoml/_internal/io_descriptors/multipart.py [hunk 3]
```diff
@@ -44,28 +59,30 @@ def predict(img, annotations):
     curl -i -F img=@test.jpg -F annotations=@test.json localhost:5000/predict
     """
 
-    def __init__(self, **inputs: IODescriptor[_DescriptorType]):
+    def __init__(self, **inputs: _DescriptorType):
         for descriptor in inputs.values():
-            if isinstance(descriptor, Multipart):
+            if isinstance(descriptor, Multipart):  # pragma: no cover
```
Comment (sauyon) [inline bentoml/_internal/io_descriptors/multipart.py:64]: Why is this no cover?

File: bentoml/_internal/io_descriptors/multipart.py [hunk 4]
```diff
@@ -44,28 +59,30 @@ def predict(img, annotations):
...
                 raise InvalidArgument(
                     "Multipart IO can not contain nested Multipart item"
                 )
 
-        self._inputs = inputs
+        self._inputs = inputs  # type: t.Dict[str, _DescriptorType]
```
Comment (sauyon) [inline bentoml/_internal/io_descriptors/multipart.py:69]: can you use `self._inputs: TYPE = inputs` here?

File: bentoml/_internal/io_descriptors/multipart.py [hunk 5]
```diff
@@ -83,7 +100,41 @@ async def from_http_request(self, request: Request) -> MultipartIO:
             res[k] = v
         return res
 
-    async def to_http_response(self, obj: MultipartIO) -> Response:
+    @t.overload
+    async def to_http_response(self, obj: t.Dict[str, str]) -> Response:  # noqa: F811
```
Comment (sauyon) [inline bentoml/_internal/io_descriptors/multipart.py:104]: Seems like pyflakes is supposed to detect overload now---maybe it has to not be qualified?

And what's the point of this, anyway? Why not just have a union type? If there isn't separate documentation I don't think it's very valuable.

File: bentoml/_internal/io_descriptors/numpy.py [hunk 6]
```diff
@@ -22,7 +22,7 @@
 def _is_matched_shape(
     left: t.Optional[t.Tuple[int, ...]],
     right: t.Optional[t.Tuple[int, ...]],
-) -> bool:
+) -> bool:  # pragma: no cover
```
Comment (sauyon) [inline bentoml/_internal/io_descriptors/numpy.py:25]: Why no cover?

File: bentoml/_internal/io_descriptors/pandas.py [hunk 7]
```diff
@@ -16,10 +16,46 @@
 else:
     pd = LazyLoader("pd", globals(), "pandas")
 
+try:
+    from typing import Literal
+except ImportError:  # pragma: no cover
```
Comment (sauyon) [inline bentoml/_internal/io_descriptors/pandas.py:21]: Does this not get covered by our 3.7 test?

File: bentoml/_internal/io_descriptors/pandas.py [hunk 8]
```diff
@@ -16,10 +16,46 @@
...
+    from typing_extensions import Literal
+
 logger = logging.getLogger(__name__)
 
 
-class PandasDataFrame(IODescriptor):
+def _infer_type(item: str) -> str:  # pragma: no cover
```
Comment (sauyon) [inline bentoml/_internal/io_descriptors/pandas.py:27]: seems like there are multiple infer type methods?

File: bentoml/_internal/io_descriptors/pandas.py [hunk 9]
```diff
@@ -16,10 +16,46 @@
...
+        return "string"
+    elif item.startswith("bool"):
+        return "boolean"
+    else:
+        return "object"
+
+
+def _schema_type(
+    mime_type: str, dtype: t.Optional[t.Union[bool, t.Dict[str, t.Any]]]
+) -> t.Dict[str, t.Any]:  # pragma: no cover
```
Comment (sauyon) [inline bentoml/_internal/io_descriptors/pandas.py:42]: why no cover?

File: bentoml/_internal/types.py [hunk 10]
```diff
@@ -36,7 +36,7 @@
 
 
 @total_ordering
-class Tag:
+class Tag(object):
```
Comment (sauyon) [inline bentoml/_internal/types.py:39]: Should I have done this everywhere?

File: bentoml/_internal/service/inference_api.py [hunk 11]
```diff
@@ -45,9 +45,11 @@ def __init__(
...
 
-    def __repr__(self):
-        return "<{}({} -> {})>".format(
-            self.__class__.__name__, repr(self.input), repr(self.output)
+    def __str__(self):
+        return u"%s(%s\u2192 %s)" % (
+            self.__class__.__name__,
+            str(self.input),
+            str(self.output),
         )
```
Comment (sauyon) [inline bentoml/_internal/service/inference_api.py:53]: I feel like f strings are more readable than formats like this.

Also, why not just have the actual unicode character?

**Review by sauyon — COMMENTED** [review]
A mostly stylistic review, since I still don't really know much about what this is supposed to do.

File: typings/PIL/GimpPaletteFile.pyi [hunk 12]
```diff
@@ -1,6 +1,3 @@
-"""
-This type stub file was generated by pyright.
-"""
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply typings/PIL/GimpPaletteFile.pyi:3]: cleanup

File: bentoml/_internal/io_descriptors/base.py [hunk 13]
```diff
@@ -7,6 +8,20 @@
 IOPyObj = t.TypeVar("IOPyObj")
 
 
+def _mk_str(obj: t.Any) -> str:
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/base.py:11]: I don't think so, this is mainly for IO descriptor only. I don't see other use for it

File: bentoml/_internal/io_descriptors/base.py [hunk 14]
```diff
@@ -15,10 +30,18 @@ class IODescriptor(ABC, t.Generic[IOPyObj]):
...
+        return f"%s(%s)" % (
+            self.__class__.__name__,
+            ", ".join(
+                [f'{k.strip("_")}={_mk_str(v)}' for k, v in self.__dict__.items()]
+            ),
+        )
 
     # fmt: off
+    @abstractmethod
+    def openapi_schema_type(self) -> t.Dict[str, str]: ...  # noqa: E704
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/base.py:43]: we should enable fmt: on and remove E704 ignore. I forgot why I did this :)

File: bentoml/_internal/io_descriptors/file.py [hunk 15]
```diff
@@ -54,20 +54,19 @@ def predict(input_pdf):
         IO Descriptor that represents file-like objects.
     """  # noqa: LN001
 
-    def __init__(self, media_type: t.Optional[str] = None):
-        self._media_type = media_type or "application/octet-stream"
+    def __init__(self, mime_type: t.Optional[str] = None):
+        self._mime_type = mime_type or "application/octet-stream"
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/file.py:58]: maybe we should just do 
```python
self._mime_type = 'application/octet-stream' if mime_type is None else mime_type
```

File: bentoml/_internal/io_descriptors/file.py [hunk 16]
```diff
@@ -54,20 +54,19 @@ def predict(input_pdf):
...
 
-    def openapi_schema(self) -> t.Dict[str, t.Dict[str, t.Any]]:
-        # TODO: supports multipart OpenAPI
-        return {self._media_type: dict(schema=dict(type="string", format="binary"))}
+    def openapi_schema_type(self) -> t.Dict[str, str]:
+        return dict(type="string", format="binary")
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/file.py:61]: sure, I don't really have a preference

File: bentoml/_internal/io_descriptors/multipart.py (hunk 4 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/multipart.py:69]: sure I will add it tmr morning

File: bentoml/_internal/io_descriptors/multipart.py (hunk 5 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/multipart.py:104]: Union type doesn't seem to work with pyright

File: bentoml/_internal/io_descriptors/numpy.py (hunk 6 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/numpy.py:25]: I don't think we need to write unit test for this

File: bentoml/_internal/io_descriptors/pandas.py (hunk 8 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/pandas.py:27]: only for pandas and numpy

File: bentoml/_internal/io_descriptors/pandas.py (hunk 7 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/pandas.py:21]: yep

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/pandas.py:21]: well usually I don't think we need to write test to cover import

File: bentoml/_internal/service/inference_api.py (hunk 11 above)
Comment (aarnphm) [reply bentoml/_internal/service/inference_api.py:53]: sure I will change it to f-string

File: bentoml/_internal/io_descriptors/multipart.py (hunk 5 above)
Comment (sauyon) [reply bentoml/_internal/io_descriptors/multipart.py:104]: I haven't seen an issue with that? What was wrong with it?

File: bentoml/_internal/io_descriptors/numpy.py (hunk 6 above)
Comment (sauyon) [reply bentoml/_internal/io_descriptors/numpy.py:25]: Are we even using it? What's it for?

File: bentoml/_internal/io_descriptors/pandas.py (hunk 7 above)
Comment (sauyon) [reply bentoml/_internal/io_descriptors/pandas.py:21]: no, but in this particular case shouldn't it be covered?

File: bentoml/_internal/io_descriptors/pandas.py (hunk 8 above)
Comment (sauyon) [reply bentoml/_internal/io_descriptors/pandas.py:27]: should those be unified or do they need to be distinct?

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/pandas.py:27]: distinct

File: bentoml/_internal/types.py (hunk 10 above)
Comment (aarnphm) [reply bentoml/_internal/types.py:39]: pyright was angry b4 that

File: bentoml/_internal/io_descriptors/pandas.py (hunk 9 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/pandas.py:42]: technically this should be covered in the tests, but it didn't for some reason

File: bentoml/_internal/io_descriptors/multipart.py [hunk 17]
```diff
@@ -83,7 +144,35 @@ async def from_http_request(self, request: Request) -> MultipartIO:
...
+    async def to_http_response(
+        self, obj: t.Dict[str, "np.ndarray[t.Any, np.dtype[t.Any]]"]
+    ) -> Response:
+        ...
+
+    @t.overload
+    async def to_http_response(self, obj: t.Dict[str, "FileLike"]) -> Response:
+        ...
+
+    @t.overload
```
Comment (bojiang) [inline bentoml/_internal/io_descriptors/multipart.py:169]: we should have just Union here rather than overloads?

File: bentoml/_internal/io_descriptors/base.py [hunk 18]
```diff
@@ -1,12 +1,52 @@
...
+from ..types import FileLike
+
+if TYPE_CHECKING:
+    import numpy as np
+    import pandas as pd
+    import PIL.Image
+    from pydantic import BaseModel
+
+
+JSONType = t.Union[str, t.Dict[str, t.Any], "BaseModel"]
```
Comment (bojiang) [inline bentoml/_internal/io_descriptors/base.py:18]: BaseModel?

File: bentoml/_internal/io_descriptors/multipart.py [hunk 19]
```diff
@@ -10,62 +10,123 @@
...
+    Args:
+        inputs (`Dict[str, IODescriptor]`):
+            Dictionary consisting keys as inputs definition for a Multipart request/response, values
+             as IODescriptor supported by BentoML. Currently we support Image, NumpyNdarray, PandasDataFrame,
+             PandasSeries, Text, File.
+
+    .. notes::
+        Make sure to match your input params in your API function to the keys defined under `Multipart`::
+
+            ┌───────────────────────────────────────────────────────┐
```
Comment (bojiang) [inline bentoml/_internal/io_descriptors/multipart.py:89]: 👍

File: bentoml/_internal/io_descriptors/base.py (hunk 18 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/base.py:21]: oops my bad forgot to remove it

File: bentoml/_internal/io_descriptors/multipart.py (hunk 17 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/multipart.py:170]: for some reason pyright make a false negative with Union type

File: bentoml/_internal/io_descriptors/multipart.py (hunk 3 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/multipart.py:114]: actually when writing code if users do this then they already show it to them

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/multipart.py:114]: no need to cover since type checker should detect this when users write their code. This acts as safety net

File: bentoml/_internal/io_descriptors/base.py [hunk 20]
```diff
@@ -1,44 +1,116 @@
...
+            if len(args) != 0:
+                default_params.update(zip_longest(args, defaults))
+            if spec.varkw is not None:
+                default_params[spec.varkw] = spec.kwonlydefaults
+            setattr(self, "__io_default_params__", default_params)
 
-    def __repr__(self) -> str:
-        return f"{self.__class__.__name__}"
+    def __str__(self) -> str:
+        default_params = getattr(self, "__io_default_params__")
```
Comment (bojiang) [inline bentoml/_internal/io_descriptors/base.py:73]: I guess u want `getattr(self, "__io_default_params__", None)`?
The current will raise the same exception if `__io_default_params__ ` doesn't exist

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/base.py]: fixed

File: bentoml/_internal/io_descriptors/base.py [hunk 21]
```diff
@@ -1,44 +1,116 @@
...
+            if spec.varkw is not None:
+                default_params[spec.varkw] = spec.kwonlydefaults
+            setattr(self, "__io_default_params__", default_params)
 
-    def __repr__(self) -> str:
-        return f"{self.__class__.__name__}"
+    def __str__(self) -> str:
+        default_params = getattr(self, "__io_default_params__", None)
+        filtered: t.List[str] = []
+        for k, v in self.__dict__.items():
```
Comment (bojiang) [inline bentoml/_internal/io_descriptors/base.py:75]: it might be simpler if implemented in __new__

File: bentoml/_internal/io_descriptors/json.py [hunk 22]
```diff
@@ -42,7 +41,7 @@ def default(self, o: _SerializableObj) -> t.Any:
         if isinstance(o, np.ndarray):
             return o.tolist()
 
-        if any(isinstance(o, i) for i in [pd.DataFrame, pd.Series]):
+        if isinstance(o, (pd.DataFrame, pd.Series)):
             raise BentoMLException(
```
Comment (parano) [inline bentoml/_internal/io_descriptors/json.py:45]: Just noticed this exception -why did we make this change? why not just try converting it to JSON here? it is quite common to return a dictionary that contains dataframe and it would be nice to just support that - PandasDataFrame doesn't really handle that case neither

File: bentoml/_internal/io_descriptors/json.py (hunk 2 above)
Comment (parano) [reply bentoml/_internal/io_descriptors/json.py:147]: maybe we should consider returning the pydantic obj instead of json_obj here? or make that an option?  let's have a follow up discussion

File: bentoml/_internal/io_descriptors/multipart.py [hunk 23]
```diff
@@ -10,62 +10,140 @@
...
+        input_spec = Multipart(arr=NumpyNdarray(), annotations=JSON())
+        output_spec = Multipart(output=NumpyNdarray(), result=JSON())
+
+        @svc.api(input=input_spec, output=output_spec)
+        def predict(arr, annotations):
+            res = runner.run(arr)
+            return {"output":res, "result":annotations}
+
+    Users then can then serve this service with `bentoml serve`::
+        % bentoml serve ./sklearn_svc.py:svc --auto-reload
```
Comment (parano) [inline bentoml/_internal/io_descriptors/multipart.py:51]: `s/--auto-reload/--reload/`

File: bentoml/_internal/io_descriptors/json.py (hunk 22 above)
Comment (bojiang) [reply bentoml/_internal/io_descriptors/json.py:44]: how about
```
@api(output=Multipart(df=DataFrame()))
def api():
    return {"df": df}

```

Comment (bojiang) [reply bentoml/_internal/io_descriptors/json.py:44]: shall we encourage this usage since we already supported

```
@api(output=Multipart(df=DataFrame()))
def api():
    return {"df": df}

```
?

File: bentoml/_internal/io_descriptors/multipart.py (hunk 23 above)
Comment (bojiang) [reply bentoml/_internal/io_descriptors/multipart.py:51]:
```suggestion
        % bentoml serve ./sklearn_svc.py:svc --reload
```

**Review by bojiang — APPROVED** [review]

File: bentoml/_internal/io_descriptors/json.py (hunk 22 above)
Comment (parano) [reply bentoml/_internal/io_descriptors/json.py:44]: it's quite different, right? one is a multi-part response vs. a simple JSON response

File: bentoml/_internal/io_descriptors/base.py [hunk 24]
```diff
@@ -1,44 +1,86 @@
...
 
     HTTP_METHODS = ["POST"]
+    _init_str: str = ""
+
+    def __new__(cls: t.Type[_T], *args: t.Any, **kwargs: t.Any) -> _T:
+        self = super().__new__(cls)
+        arg_strs = tuple(repr(i) for i in args) + tuple(
+            f'{k}={repr(v)}' for k, v in kwargs.items()
+        )
+        setattr(self, "_init_str", f"{cls.__name__}({', '.join(arg_strs)})")
```
Comment (parano) [inline bentoml/_internal/io_descriptors/base.py:51]: should we do anything for IO descriptors that are created via `from_sample`?

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/base.py:51]: maybe add a attribute `_from_sample` = True?

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/base.py:51]: we can save it as an example, the present it on openapi as a $ref to that example

File: tests/integration/conftest.py [hunk 25]
```diff
@@ -21,10 +40,61 @@ def pytest_collection_modifyitems(config, items):
...
+def fixture_modelstore(tmp_path_factory: "TempPathFactory") -> "ModelStore":
     # we need to get consistent cache folder, thus tmpdir is not usable here
     # NOTE: after using modelstore, also use `delete_cache_model` to remove model after
     #  load tests.
     path = tmp_path_factory.mktemp("bentoml")
     return ModelStore(path)
+
+
+@pytest.fixture(scope="session", name="pipeline")
+def fixture_pipeline(modelstore: "ModelStore") -> Pipeline:
```
Comment (parano) [inline tests/integration/conftest.py:53]: this seems way too much complexity for something that can be done in a few lines of code, can we just use a utility function instead?

File: tests/integration/frameworks/test_catboost_impl.py [hunk 26]
```diff
@@ -1,23 +1,23 @@
...
-if t.TYPE_CHECKING:
-    from bentoml._internal.models import ModelStore
+if TYPE_CHECKING:
+    from bentoml._internal.models import Model, ModelStore
 
-TEST_MODEL_NAME = __name__.split(".")[-1]
 
-
-def catboost_model() -> cbt.core.CatBoostClassifier:
+def create_catboost_model() -> cbt.core.CatBoostClassifier:
```
Comment (parano) [inline tests/integration/frameworks/test_catboost_impl.py:20]: let's modify this to `create_catboost_model(save_params, metadata)` and replace the `pipleline` fixture?

File: tests/integration/conftest.py (hunk 25 above)
Comment (parano) [reply tests/integration/conftest.py:53]: I think in general fixtures are better fitted for initializing service/object and states. For a reusable "process", simple function calls are easier to manage

File: bentoml/detectron.py [hunk 27]
```diff
@@ -59,18 +67,18 @@ def load(
             "my_detectron_model:20201012_DE43A2")
     """  # noqa
 
-    model = model_store.get(tag)
```
Comment (parano) [inline bentoml/detectron.py:62]: what's the reason for this chagne?

File: bentoml/detectron.py [hunk 28]
```diff
@@ -208,33 +219,44 @@ def num_replica(self) -> int:
...
             device = "cpu"
         self._model = load(self._tag, device, self._model_store)
         self._predict_fn = getattr(self._model, self._predict_fn_name)
 
-    # pylint: disable=arguments-differ
-    def _run_batch(  # type: ignore[override]
-        self, input_data: t.Union[np.ndarray, torch.Tensor]
-    ) -> np.ndarray:
-        images = np.split(input_data, input_data.shape[0], 0)
+    def _run_batch(
```
Comment (parano) [inline bentoml/detectron.py:231]: should we move this to a separate PR?

File: bentoml/sklearn.py [hunk 29]
```diff
@@ -137,31 +147,32 @@ def __init__(
...
     def num_concurrency_per_replica(self) -> int:
-        # NOTE: sklearn doesn't use GPU, so return max. no. of CPU's.
-        return int(round(self.resource_quota.cpu))
+        return 1
 
     @property
     def num_replica(self) -> int:
-        # NOTE: sklearn doesn't use GPU, so just return 1.
-        return 1
+        return int(round(self.resource_quota.cpu))
```
Comment (parano) [inline bentoml/sklearn.py:154]: with `parallel_backend`, this should just return 1

File: bentoml/sklearn.py [hunk 30]
```diff
@@ -137,31 +147,32 @@ def __init__(
 
     @property
     def num_concurrency_per_replica(self) -> int:
-        # NOTE: sklearn doesn't use GPU, so return max. no. of CPU's.
-        return int(round(self.resource_quota.cpu))
+        return 1
```
Comment (parano) [inline bentoml/sklearn.py:150]: also need to change this back

File: bentoml/sklearn.py (hunk 29 above)
Comment (parano) [reply bentoml/sklearn.py:154]: for this kind of changes it is better to put in separate PR for visibility

File: bentoml/sklearn.py [hunk 31]
```diff
@@ -137,31 +147,32 @@ def __init__(
...
     @property
     def required_models(self) -> t.List[Tag]:
         return [self._model_info.tag]
 
-    # pylint: disable=arguments-differ,attribute-defined-outside-init
-    def _setup(self) -> None:  # type: ignore[override]
+    # pylint: disable=attribute-defined-outside-init
+    def _setup(self) -> None:
         self._model = joblib.load(filename=self._model_file)
+        self._infer_func = getattr(self._model, self._function_name)
```
Comment (parano) [inline bentoml/sklearn.py:163]: move this to a separate class method? e.g.:

```python
def _infer(self, *args, **kwargs):
    with self._parallel_ctx:
          return getattr(self._model, self._function_name)(*args, **kwargs)
```

File: requirements/dev-requirements.txt [hunk 32]
```diff
@@ -5,6 +5,7 @@ setuptools
 twine
 wheel
 pydantic
+attrs>=21.1
```
Comment (parano) [inline requirements/dev-requirements.txt:8]: duplicated below in line 13

File: scripts/Dockerfile-checker [hunk 33]
```diff
@@ -36,6 +36,10 @@ RUN npm i -g npm@^7 yarn pyright
 
 RUN python -m pip install -U pip setuptools
 
+ENV VIRTUAL_ENV=/opt/venv
```
Comment (parano) [inline scripts/Dockerfile-checker:39]: I thought we are getting rid of the checker image? is it still being used?

File: scripts/ci/config.yml [hunk 34]
```diff
@@ -24,6 +24,7 @@ catboost:
   <<: *tmpl
   dependencies:
     - "catboost"
+    - "psutil"
```
Comment (parano) [inline scripts/ci/config.yml:27]: already a bentoml dependency

Comment (parano) [reply scripts/ci/config.yml:27]:
```suggestion
```

File: requirements/dev-requirements.txt (hunk 32 above)
Comment (parano) [reply requirements/dev-requirements.txt:8]:
```suggestion
```

File: bentoml/sklearn.py [hunk 35]
```diff
@@ -137,31 +147,32 @@ def __init__(
```
Comment (parano) [inline bentoml/sklearn.py:147]: we should also change the parallel_backend mode above, from 'threading' to 'loky': https://scikit-learn.org/stable/modules/generated/sklearn.utils.parallel_backend.html

File: bentoml/_internal/io_descriptors/base.py (hunk 24 above)
Comment (parano) [reply bentoml/_internal/io_descriptors/base.py:51]: that would be really nice

File: bentoml/_internal/io_descriptors/json.py [hunk 36]
```diff
@@ -42,7 +40,7 @@ def default(self, o: _SerializableObj) -> t.Any:
         if isinstance(o, np.ndarray):
             return o.tolist()
 
-        if any(isinstance(o, i) for i in [pd.DataFrame, pd.Series]):
+        if isinstance(o, (pd.DataFrame, pd.Series)):
```
Comment (parano) [inline bentoml/_internal/io_descriptors/json.py:43]:
```
if pd is not None:
           if isinstance(o, (pd.DataFrame, pd.Series)):
                       return o.to_json(orient=default_orient)
```

File: bentoml/_internal/io_descriptors/json.py [hunk 37]
```diff
@@ -42,7 +40,7 @@ def default(self, o: _SerializableObj) -> t.Any:
...
             raise BentoMLException(
```
Comment (parano) [inline bentoml/_internal/io_descriptors/json.py:44]: should we also add support for returning a pydantic model here?

File: bentoml/_internal/io_descriptors/json.py (hunk 2 above)
Comment (parano) [reply bentoml/_internal/io_descriptors/json.py:141]: just discussed with @bojiang, we all agree that it should just return the pydantic object in this case, cc @aarnphm

File: bentoml/_internal/io_descriptors/json.py (hunk 37 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/json.py:44]: I don't think that is the job of a serializer. we should just convert a pydantic model to json

File: bentoml/_internal/io_descriptors/json.py (hunk 36 above)
Comment (aarnphm) [reply bentoml/_internal/io_descriptors/json.py:43]: where do we defin this default_orient?

File: bentoml/_internal/io_descriptors/json.py (hunk 37 above)
Comment (parano) [reply bentoml/_internal/io_descriptors/json.py]: yes that's that I meant, if it's a pydantic instance here, convert it to dict

Comment (aarnphm) [reply bentoml/_internal/io_descriptors/json.py]: it should do it now

File: bentoml/_internal/io_descriptors/pandas.py [hunk 38]
```diff
@@ -124,6 +160,9 @@ def __init__(
         enforce_dtype: bool = False,
         shape: t.Optional[t.Tuple[int, ...]] = None,
         enforce_shape: bool = False,
+        mime_type: t.Union[
+            Literal["text/csv"], Literal["application/json"]
```
Comment (parano) [inline bentoml/_internal/io_descriptors/pandas.py:164]: "text/csv" doesn't work with `read_json` call below right?

File: bentoml/_internal/io_descriptors/pandas.py [hunk 39]
```diff
@@ -151,6 +196,18 @@ async def from_http_request(self, request: Request) -> "pd.DataFrame":
...
+        if mime_type not in ["application/json", "text/csv"]:
+            raise BentoMLException(
+                f"mimetype {mime_type} is not supported, BentoML will only"
+                " supports `application/json` or `text/csv`"
+            )
+        if mime_type != self._mime_type:
+            logger.warning(
+                f"Current mimetype is {self._mime_type}, while receiving mimetype"
+                f" to be {mime_type}. This will lead to mismatch OpenAPI specification."
+            )
```
Comment (parano) [inline bentoml/_internal/io_descriptors/pandas.py:210]: if we were to support `text/csv`, we also need to change l219 below

File: bentoml/_internal/io_descriptors/pandas.py [hunk 40]
```diff
@@ -368,18 +425,29 @@ def __init__(
...
+    def openapi_request_schema(self) -> t.Dict[str, t.Any]:
+        """Returns OpenAPI schema for incoming requests"""
+        return {self._mime_type: {"schema": self.openapi_schema_type()}}
+
+    def openapi_responses_schema(self) -> t.Dict[str, t.Any]:
+        """Returns OpenAPI schema for outcoming responses"""
+        return {self._mime_type: {"schema": self.openapi_schema_type()}}
 
-    async def from_http_request(self, request: Request) -> "pd.Series":
+    async def from_http_request(self, request: Request) -> "pd.Series[t.Any]":
```
Comment (parano) [inline bentoml/_internal/io_descriptors/pandas.py:450]: same here in L470

File: bentoml/_internal/io_descriptors/text.py [hunk 41]
```diff
@@ -56,17 +55,23 @@ def predict(input_arr):
         IO Descriptor that represents strings type.
     """
 
-    @staticmethod
-    def openapi_schema() -> t.Dict[str, t.Dict[str, t.Any]]:
-        return {MIME_TYPE: {"schema": {"type": "string"}}}
+    def __init__(self, *args: t.Any, **kwargs: t.Any) -> None:
+        if len(args) > 0 or len(kwargs) > 0:
+            raise EnvironmentError(
```
Comment (parano) [inline bentoml/_internal/io_descriptors/text.py:60]: this should be a TypeError?

Comment (parano) [reply bentoml/_internal/io_descriptors/text.py:60]: can we just remove the `__init__` here?

File: bentoml/_internal/io_descriptors/pandas.py (hunk 40 above)
Comment (parano) [reply bentoml/_internal/io_descriptors/pandas.py:450]: we can do that in a follow up PR

File: bentoml/_internal/io_descriptors/text.py [hunk 42]
```diff
@@ -56,17 +55,23 @@ def predict(input_arr):
...
+                f"{self.__class__.__name__} is not designed to take any arguments."
+                " You should only call `Text()` instead."
+            )
+
```
Comment (parano) [inline bentoml/_internal/io_descriptors/text.py:64]:
```suggestion
```
